netconf.c

```diff
@@ -853,8 +864,511 @@ get_full_tree ()
+        xmlFreeNode (xml);
+        xml = NULL;
+    }
+
+    xmlXPathFreeObject(xpath_obj);
+    xmlXPathFreeContext(xpath_ctx);
+
+
+    /* Cleaning up a doc is tricky */
+    if (xml)
+        xmlSetTreeDoc (xmlDocGetRootElement(doc), NULL);
+
+    root_node = xmlNewNode (NULL, BAD_CAST "root");
+    xmlDocSetRootElement (doc, root_node);
+    xmlFreeDoc (doc);
+
+    return status == 0;
+}
+
+static char *
+find_first_non_path(char *path)
+{
+    char *ptr = path;
+    bool slash = false;
+    int len = strlen (path);
+    int i;
+    for (i = 0; i < len; i++)
+    {
+        if (*ptr == '*' || *ptr == '[' || *ptr == '@')
+        {
+            if (slash)
+                ptr--;
+
+            return ptr;
+        }
+        if (*ptr == '/')
+        {
+            if (slash)
+                return ptr - 1;
+
+            slash = true;
+        }
+        else
+            slash = false;
+        ptr++;
+    }
+    return NULL;
+}
+
+static xpath_type
+prepare_xpath_query_path (char *path, char *schema_path, char **sch_path)
+{
+    char *non_path = NULL;
+    int len;
+
+    *sch_path = NULL;
+    if (path[0] != '/')
+    {
+        return XPATH_ERROR;
+    }
+    len = strlen (path);
+    if (len == 1)
+    {
+        *sch_path = g_strdup (path);
+        return XPATH_SIMPLE;
+    }
+
+    /* Check for // syntax */
+    if (path[1] == '/')
+    {
+        if (schema_path)
+        {
+            *sch_path = g_strdup ((char *) schema_path);
+            return XPATH_EVALUATE;
+        }
+        /* Trying to do a // query but we have no namespace. This will not work */
+        return XPATH_ERROR;
+    }
+    non_path = find_first_non_path (path);
+    if (non_path)
+    {
+        *sch_path = g_strndup (path, non_path - path);
+        return XPATH_EVALUATE;
+    }
+    *sch_path = g_strdup (path);
+    return XPATH_SIMPLE;
+}
+
+static char *
+find_model_path_by_href_prefix (char *href, char *prefix)
+{
+    xmlNode *xml = sch_child_first (g_schema);
+    char *path;
+    char *name;
+
+    while (xml && xml->type == XML_ELEMENT_NODE)
+    {
+        if (xml->ns &&
+            ((href && xml->ns->href && g_strcmp0 ((char *) xml->ns->href, href) == 0) ||
+             (!href && xml->ns->prefix && g_strcmp0 ((char *) xml->ns->prefix, prefix) == 0)))
+        {
+            name = (char *) xmlGetProp (xml, (xmlChar *) "name");
+            path = g_strdup_printf ("/%s:%s", prefix, name);
+            xmlFree (name);
+            return path;
+        }
+        xml = xml->next;
+    }
+    return NULL;
+}
+
+static char *
+check_namespace_set (xmlNode *node, char **ns_href, char **ns_prefix)
+{
+    xmlNs *ns = node->nsDef;
+    char *path;
+    char *schema_path = NULL;
+
+    while (ns)
+    {
+        path = find_model_path_by_href_prefix ((char *) ns->href, (char *) ns->prefix);
+        schema_path = path;
+        if (path)
+        {
+            if (ns->href)
+            {
+                if (ns_href && *ns_href)
```

Comment:
Minor niggle here and the next if statement - you check ns_href before checking *ns_href, but then use *ns_href without the check. If ns_href can truly be NULL this would crash. If it truly can't be NULL, then no need to check it.

Reply:
This has been fixed in the latest patch